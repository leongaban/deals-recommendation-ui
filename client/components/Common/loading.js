import React from 'react';
import { withStyles } from '@material-ui/core/styles';
import CircularProgress from '@material-ui/core/CircularProgress';

const styles = theme => ({
  progress: {
    margin: theme.spacing.unit * 2
  },
  loadingWrapper: {
    display: 'flex',
    height: '100%',
    alignItems: 'center',
    justifyContent: 'center'
  }
});

function CircularIndeterminate(props) {
  const { classes } = props;
  return (
    <div className={classes.loadingWrapper}>
      <CircularProgress className={classes.progress} size={50} />
    </div>
  );
}

export default withStyles(styles)(CircularIndeterminate);
