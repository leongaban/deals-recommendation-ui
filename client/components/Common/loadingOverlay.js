import React from 'react';

// Material-ui
import CircularProgress from '@material-ui/core/CircularProgress';

const LoadingOverlay = () => (
  <div className="loading-overlay">
    <h1>UPDATING RULES.</h1>
    <CircularProgress size={250} />
  </div>);

export default LoadingOverlay;
