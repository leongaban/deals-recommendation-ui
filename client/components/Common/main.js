import React from 'react';
import { connect } from 'react-redux';

// MUI
import { withStyles } from '@material-ui/core/styles';

// Components
import TopNav from 'components/Common/topNav';
import ProductModal from 'components/Common/Modals/productModal';
import DeleteModal from 'components/Common/Modals/deleteModal';

// Utils
import { cleanMapStateToProps } from 'utils/redux';

// Strings
import {
  PRODUCT_MODAL,
  DELETE_PRODUCT_MODAL
} from 'constants/modals';

import {
  findNotification,
  showNotification
} from 'utils/notifications';

const styles = {
  root: {
    flexGrow: 1,
    display: 'flex',
    flexDirection: 'column',
    height: '100%'
  }
};

const renderModal = (modalName, modalProps) => {
  switch (modalName) {
    case PRODUCT_MODAL:
      return <ProductModal {...modalProps} />;
    case DELETE_PRODUCT_MODAL:
      return <DeleteModal {...modalProps} />;
    default:
      return null;
  }
};

const Main = (props) => {
  const {
    classes,
    children,
    notifications,
    modalName,
    modalProps,
    currentNotification
  } = props;

  const notificationObj = findNotification(currentNotification, notifications);

  return (
    <div className={classes.root}>
      { renderModal(modalName, modalProps) }
      <TopNav />
      { showNotification(notificationObj)}
      {children}
    </div>
  );
};

export const MainJest = Main;

export default connect(cleanMapStateToProps([
  'modalName',
  'modalProps'
]), null)(withStyles(styles)(Main));
