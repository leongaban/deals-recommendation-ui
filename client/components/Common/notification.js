import React from 'react';
import Snackbar from '@material-ui/core/Snackbar';

const Notification = props => (
  <Snackbar
    id="notifications-snackbar"
    className={props.className}
    open={!!props.open}
    message={props.message || ''}
    autoHideDuration={6000}
  />
);

export default Notification;
