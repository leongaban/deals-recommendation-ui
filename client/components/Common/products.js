import React, { Component } from 'react';
import { connect } from 'react-redux';

// Actions
import { setModal as setModalAction } from 'actions/Modals';

// Utils
import { cleanMapStateToProps } from 'utils/redux';

// Strings
import { PRODUCT_MODAL } from 'constants/modals';
import { ADD_A_PRODUCT } from 'copy/Components/modals';

// Components
import ContentSection from 'components/Layout/contentSection';
import TableHeader from 'components/Layout/tableHeader';
import TableContainer from 'components/Layout/tableContainer';
import LoadingOverlay from 'components/Common/loadingOverlay';

class Products extends Component {
  static getDerivedStateFromProps({ rulesLoading }) {
    return { rulesLoading };
  }

  constructor(props) {
    super(props);
    const { rulesLoading } = props;
    this.state = { rulesLoading };
  }

  render() {
    const { setModal } = this.props;
    const { rulesLoading } = this.state;

    return (
      <ContentSection>
        { rulesLoading && <LoadingOverlay /> }
        <section className="header-container">
          <header>
            <section
              className="btn-add-container"
              onClick={() => setModal(PRODUCT_MODAL)}
            >
              <div className="btn-add-product">
                <div className="circle" />
                <img src="/static/imgs/icon-add.png" alt={ADD_A_PRODUCT} />
              </div>
              <p>{ADD_A_PRODUCT}</p>
            </section>
          </header>
        </section>
        <section className="table-container">
          <TableHeader />
          <TableContainer />
        </section>
      </ContentSection>
    );
  }
}

export const mapDispatchToProps = dispatch => ({
  setModal: (...args) => dispatch(setModalAction(...args))
});

const mapStateToProps = cleanMapStateToProps([
  'rulesLoading'
]);

export const ProductsJest = Products;

export default connect(mapStateToProps, mapDispatchToProps)(Products);
