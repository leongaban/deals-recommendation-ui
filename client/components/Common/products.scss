.header-container {
  header { width: 100%; }
}

.btn-add-container {
  position: relative;
  width: 160px;
  height: 48px;
  cursor: pointer;

  p {
    position: absolute;
    top: 14px;
    margin-left: 50px;
    font-weight: 400;
    color: #000;
    text-align: right;
  }

  &:hover {
    .btn-add-product {
      background: black;
      .circle { border: 3px solid $gray; }
    }
  }
}

.btn-add-product {
  @include transition(background, .3s, ease-in-out);
  position: relative;
  width: 48px;
  height: 48px;
  background: $themeRed;
}

.circle {
  @include transition(border, .3s, ease-in-out);
  position: absolute;
  top: 8px;
  left: 8px;
  width: 32px;
  height: 32px;
  border-radius: 32px;
  -moz-border-radius: 32px;
  -webkit-border-radius: 32px;
  border: 3px solid $themeRed;
  background-color: rgba(249, 249, 249, 0);
}

.loading-overlay {
  position: fixed;
  top: 0;
  left: 0;
  z-index: 10;
  padding-top: 111px;
  width: 100%;
  height: 100%;
  background-color: rgba(237, 237, 237, 0.9);

  h1 {
    position: relative;
    z-index: 9;
    top: 127px;
    font-weight: 700;
    font-size: em(20);
    color: $themeRed;
  }
}
