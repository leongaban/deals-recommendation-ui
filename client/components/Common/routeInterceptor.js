import React from 'react';
import { connect } from 'react-redux';

// Redirect
import { Redirect } from 'react-router-dom';

// Utils
import { hasRights } from 'utils/auth';
import { cleanMapStateToProps } from 'utils/redux';

// Components
import Products from 'components/Common/products';
import Loading from 'components/Common/loading';

// Actions
import { saveRedirect } from 'actions/Global';

// Constants
import {
  PAGE_AUTH_REQS,
  ROUTEPATH_DEFAULT_PAGE,
  ROUTEPATH_LOGIN
} from 'copy/Global/routes';

// Components
import { Forbidden, NotFound } from 'components';

// Utils
import { getRouteFromPathName } from 'utils/routes';

const { AUTHORIZED } = PAGE_AUTH_REQS;

class RouteInterceptor extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      forbidden: false,
      notFound: false
    };
  }

  componentDidMount() {
    this.aclCheck(this.props);
  }

  componentWillReceiveProps(nextProps) {
    this.aclCheck(nextProps);
  }

  authRouteCheck(isAuthRoute) {
    const { userServices } = this.props;

    const isAuthorized = userServices && hasRights(userServices);
    // @TODO: userServices is always empty because AUTH_LOGGED_OUT clears it!
    // console.log('userServices', userServices)

    if (isAuthRoute) {
      // @TODO: In the future, we will check for Authorization privileges.
      this.setState({
        forbidden: !isAuthorized,
        notFound: false
      });
    } else {
      this.setState({ notFound: true });
    }
  }

  aclCheck(props) {
    const { authed, location = {} } = props;
    const { pathname } = location;

    // No user yet (Firebase still initializing).
    if (authed === null) return;

    // We don't want to redirect the user to the Login page after a successful login,
    // so we default to the root view.
    const redirectUrl = pathname === ROUTEPATH_LOGIN ? ROUTEPATH_DEFAULT_PAGE : pathname;

    const isAuthorizedRoute = getRouteFromPathName(redirectUrl, r => r.authReq === AUTHORIZED);

    // If the user is logged in, determine if they have access to this page,
    // or if it is indeed a 404.
    this.authRouteCheck(isAuthorizedRoute);

    if (authed === true) return;

    // Save the intended redirect route, if it exists.
    if (props.saveRedirect) {
      props.saveRedirect(isAuthorizedRoute ? redirectUrl : ROUTEPATH_DEFAULT_PAGE);
    }
  }

  render() {
    const { forbidden, notFound } = this.state;
    const { authed } = this.props;

    // If authentication is complete, but the user is not logged in,
    // redirect to the login view.
    if (authed === false) return <Redirect to="/login" />;

    if (forbidden) return <Forbidden />;
    if (notFound) return <NotFound />;

    // If user is logged in and accesses an unathenticated view,
    // redirect to the Products view.
    if (authed === true) return <Products />;

    return <Loading />;
  }
}

export const mapDispatchToProps = dispatch => ({
  saveRedirect: (...args) => dispatch(saveRedirect(...args))
});

export const RouteInterceptorJest = RouteInterceptor;

export default connect(cleanMapStateToProps([
  'authed',
  'location',
  'userServices'
]), mapDispatchToProps)(RouteInterceptor);
