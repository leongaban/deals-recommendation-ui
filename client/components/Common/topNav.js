import React, { Component } from 'react';
import { connect } from 'react-redux';

// MUI
import { withStyles } from '@material-ui/core/styles';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import Typography from '@material-ui/core/Typography';

// Copy
import { APP_NAME } from 'copy/Global/common';

// Components
import RightMenu from 'components/Menus/rightMenu';

// Utils
import { cleanMapStateToProps } from 'utils/redux';

const styles = {
  flex: {
    flex: 1
  },
  menuButton: {
    marginLeft: -12,
    marginRight: 20
  }
};

class TopNav extends Component {
  /**
   * Displays the user name (if the user is logged in).
   * @return {React.Component}
   */
  renderUserName() {
    const { authed, userFirstName, userLastName } = this.props;
    return authed && <span>{userFirstName} {userLastName}</span>;
  }

  render() {
    const { classes } = this.props;

    return (
      <AppBar position="static">
        <Toolbar>
          <Typography
            className={classes.flex}
            variant="title"
            color="inherit"
          >
            {APP_NAME}
          </Typography>
          { this.renderUserName() }
          <RightMenu />
        </Toolbar>
      </AppBar>
    );
  }
}

const stylesHOC = withStyles(styles)(TopNav);

export const TopNavJest = TopNav;

export default connect(cleanMapStateToProps([
  'authed',
  'userFirstName',
  'userLastName'
]), null)(stylesHOC);
