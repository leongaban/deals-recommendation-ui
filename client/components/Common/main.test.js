// Test utils
import { testCommonComponentAttrs } from 'utils/tests';

// Components
import { MainJest } from 'components/Common/main';

const mockProps = {
  classes: {
    root: {
      flexGrow: 1,
      display: 'flex',
      flexDirection: 'column',
      height: '100%'
    }
  }
};

describe('<MainJest /> component', () => {
  testCommonComponentAttrs(MainJest, mockProps);
});
