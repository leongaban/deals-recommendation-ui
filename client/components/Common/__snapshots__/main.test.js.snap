// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MainJest /> component when rendering should render a component matching the snapshot 1`] = `
<div
  className={
    Object {
      "display": "flex",
      "flexDirection": "column",
      "flexGrow": 1,
      "height": "100%",
    }
  }
>
  <Connect(WithStyles(TopNav)) />
</div>
`;
